File: binding/src/libqcompleter.cpp
#include <QAbstractItemModel>
#include <QAbstractItemView>
#include <QCompleter>
#include <QEvent>
#include <QModelIndex>
#include <QObject>
#include <QRect>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QWidget>
#include <qcompleter.h>
#include "libqcompleter.h"
#include "libqcompleter.hxx"

QCompleter* QCompleter_new() {
	 return new VirtualQCompleter();
}

QCompleter* QCompleter_new2(QAbstractItemModel* model) {
	 return new VirtualQCompleter(model);
}

QCompleter* QCompleter_new3(const libqt_list completions) {
	 return new VirtualQCompleter(QList<QString>());
}

QCompleter* QCompleter_new4(QObject* parent) {
	 return new VirtualQCompleter(parent);
}

QCompleter* QCompleter_new5(QAbstractItemModel* model, QObject* parent) {
	 return new VirtualQCompleter(model, parent);
}

QCompleter* QCompleter_new6(const libqt_list completions, QObject* parent) {
	 return new VirtualQCompleter(QList<QString>(), parent);
}

libqt_string QCompleter_Tr(const char* s) {
	QString _ret = QCompleter::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QCompleter_SetWidget(QCompleter* self, QWidget* widget) {
	self->setWidget(widget);
}

QWidget* QCompleter_Widget(const QCompleter* self) {
	return self->widget();
}

void QCompleter_SetModel(QCompleter* self, QAbstractItemModel* c) {
	self->setModel(c);
}

QAbstractItemModel* QCompleter_Model(const QCompleter* self) {
	return self->model();
}

void QCompleter_SetCompletionMode(QCompleter* self, int mode) {
	self->setCompletionMode(static_cast<QCompleter::CompletionMode>(mode));
}

int QCompleter_CompletionMode(const QCompleter* self) {
	return self->completionMode();
}

void QCompleter_SetFilterMode(QCompleter* self, int filterMode) {
	self->setFilterMode(static_cast<QFlags<Qt::MatchFlag>>(filterMode));
}

int QCompleter_FilterMode(const QCompleter* self) {
	return self->filterMode();
}

QAbstractItemView* QCompleter_Popup(const QCompleter* self) {
	return self->popup();
}

void QCompleter_SetPopup(QCompleter* self, QAbstractItemView* popup) {
	self->setPopup(popup);
}

void QCompleter_SetCaseSensitivity(QCompleter* self, int caseSensitivity) {
	self->setCaseSensitivity(static_cast<Qt::CaseSensitivity>(caseSensitivity));
}

int QCompleter_CaseSensitivity(const QCompleter* self) {
	return self->caseSensitivity();
}

void QCompleter_SetModelSorting(QCompleter* self, int sorting) {
	self->setModelSorting(static_cast<QCompleter::ModelSorting>(sorting));
}

int QCompleter_ModelSorting(const QCompleter* self) {
	return self->modelSorting();
}

void QCompleter_SetCompletionColumn(QCompleter* self, int column) {
	self->setCompletionColumn(column);
}

int QCompleter_CompletionColumn(const QCompleter* self) {
	return self->completionColumn();
}

void QCompleter_SetCompletionRole(QCompleter* self, int role) {
	self->setCompletionRole(role);
}

int QCompleter_CompletionRole(const QCompleter* self) {
	return self->completionRole();
}

bool QCompleter_WrapAround(const QCompleter* self) {
	return self->wrapAround();
}

int QCompleter_MaxVisibleItems(const QCompleter* self) {
	return self->maxVisibleItems();
}

void QCompleter_SetMaxVisibleItems(QCompleter* self, int maxItems) {
	self->setMaxVisibleItems(maxItems);
}

int QCompleter_CompletionCount(const QCompleter* self) {
	return self->completionCount();
}

bool QCompleter_SetCurrentRow(QCompleter* self, int row) {
	return self->setCurrentRow(row);
}

int QCompleter_CurrentRow(const QCompleter* self) {
	return self->currentRow();
}

QModelIndex* QCompleter_CurrentIndex(const QCompleter* self) {
	return new QModelIndex(self->currentIndex());
}

libqt_string QCompleter_CurrentCompletion(const QCompleter* self) {
	QString _ret = self->currentCompletion();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QAbstractItemModel* QCompleter_CompletionModel(const QCompleter* self) {
	return self->completionModel();
}

libqt_string QCompleter_CompletionPrefix(const QCompleter* self) {
	QString _ret = self->completionPrefix();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QCompleter_SetCompletionPrefix(QCompleter* self, const libqt_string prefix) {
	self->setCompletionPrefix(QString::fromUtf8(prefix.data, prefix.len));
}

void QCompleter_Complete(QCompleter* self) {
	self->complete();
}

void QCompleter_SetWrapAround(QCompleter* self, bool wrap) {
	self->setWrapAround(wrap);
}

libqt_string QCompleter_PathFromIndex(const QCompleter* self, const QModelIndex* index) {
	QString _ret = self->pathFromIndex(*index);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_list QCompleter_SplitPath(const QCompleter* self, const libqt_string path) {
	auto _ret = self->splitPath(QString::fromUtf8(path.data, path.len));
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}

void QCompleter_Activated(QCompleter* self, const libqt_string text) {
	self->activated(QString::fromUtf8(text.data, text.len));
}

void QCompleter_Activated2(QCompleter* self, const QModelIndex* index) {
	self->activated(*index);
}

void QCompleter_Highlighted(QCompleter* self, const libqt_string text) {
	self->highlighted(QString::fromUtf8(text.data, text.len));
}

void QCompleter_Highlighted2(QCompleter* self, const QModelIndex* index) {
	self->highlighted(*index);
}

libqt_string QCompleter_Tr2(const char* s, const char* c) {
	QString _ret = QCompleter::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QCompleter_Tr3(const char* s, const char* c, int n) {
	QString _ret = QCompleter::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QCompleter_Complete1(QCompleter* self, const QRect* rect) {
	self->complete(*rect);
}

// Base class handler implementation
libqt_string QCompleter_QBasePathFromIndex(const QCompleter* self, const QModelIndex* index) {
	auto* vqcompleter = dynamic_cast<const VirtualQCompleter*>(self);
	if (vqcompleter && vqcompleter->isVirtualQCompleter) {
vqcompleter->setQCompleter_PathFromIndex_IsBase(true);
	QString _ret = vqcompleter->pathFromIndex(*index);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}
}

// Auxiliary method to allow providing re-implementation
void QCompleter_OnPathFromIndex(const QCompleter* self, intptr_t slot) {
	auto* vqcompleter = dynamic_cast<const VirtualQCompleter*>(self);
	if (vqcompleter && vqcompleter->isVirtualQCompleter) {
vqcompleter->setQCompleter_PathFromIndex_Callback(reinterpret_cast<VirtualQCompleter::QCompleter_PathFromIndex_Callback>(slot));
}
}

// Base class handler implementation
libqt_list QCompleter_QBaseSplitPath(const QCompleter* self, const libqt_string path) {
	auto* vqcompleter = dynamic_cast<const VirtualQCompleter*>(self);
	if (vqcompleter && vqcompleter->isVirtualQCompleter) {
vqcompleter->setQCompleter_SplitPath_IsBase(true);
	auto _ret = vqcompleter->splitPath(QString::fromUtf8(path.data, path.len));
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		QByteArray _b = _ret[_i].toUtf8();
		libqt_string* _str = new libqt_string();
		_str->len = _b.length();
		_str->data = static_cast<const char*>(malloc(_str->len + 1));
		memcpy((void*)_str->data, _b.data(), _str->len);
		((char*)_str->data)[_str->len] = '\0';
		_data[_i] = _str;
	}
	return _arr;
}
}

// Auxiliary method to allow providing re-implementation
void QCompleter_OnSplitPath(const QCompleter* self, intptr_t slot) {
	auto* vqcompleter = dynamic_cast<const VirtualQCompleter*>(self);
	if (vqcompleter && vqcompleter->isVirtualQCompleter) {
vqcompleter->setQCompleter_SplitPath_Callback(reinterpret_cast<VirtualQCompleter::QCompleter_SplitPath_Callback>(slot));
}
}

// Derived class handler implementation
bool QCompleter_EventFilter(QCompleter* self, QObject* o, QEvent* e) {
	auto* vqcompleter = dynamic_cast<VirtualQCompleter*>(self);
	if (vqcompleter && vqcompleter->isVirtualQCompleter) {
	return vqcompleter->eventFilter(o, e);
	} else {
	return self->QCompleter::eventFilter(o, e);
}
}

// Base class handler implementation
bool QCompleter_QBaseEventFilter(QCompleter* self, QObject* o, QEvent* e) {
	auto* vqcompleter = dynamic_cast<VirtualQCompleter*>(self);
	if (vqcompleter && vqcompleter->isVirtualQCompleter) {
vqcompleter->setQCompleter_EventFilter_IsBase(true);
	return vqcompleter->eventFilter(o, e);
}
}

// Auxiliary method to allow providing re-implementation
void QCompleter_OnEventFilter(QCompleter* self, intptr_t slot) {
	auto* vqcompleter = dynamic_cast<VirtualQCompleter*>(self);
	if (vqcompleter && vqcompleter->isVirtualQCompleter) {
vqcompleter->setQCompleter_EventFilter_Callback(reinterpret_cast<VirtualQCompleter::QCompleter_EventFilter_Callback>(slot));
}
}

// Derived class handler implementation
bool QCompleter_Event(QCompleter* self, QEvent* param1) {
	auto* vqcompleter = dynamic_cast<VirtualQCompleter*>(self);
	if (vqcompleter && vqcompleter->isVirtualQCompleter) {
	return vqcompleter->event(param1);
	} else {
	return self->QCompleter::event(param1);
}
}

// Base class handler implementation
bool QCompleter_QBaseEvent(QCompleter* self, QEvent* param1) {
	auto* vqcompleter = dynamic_cast<VirtualQCompleter*>(self);
	if (vqcompleter && vqcompleter->isVirtualQCompleter) {
vqcompleter->setQCompleter_Event_IsBase(true);
	return vqcompleter->event(param1);
}
}

// Auxiliary method to allow providing re-implementation
void QCompleter_OnEvent(QCompleter* self, intptr_t slot) {
	auto* vqcompleter = dynamic_cast<VirtualQCompleter*>(self);
	if (vqcompleter && vqcompleter->isVirtualQCompleter) {
vqcompleter->setQCompleter_Event_Callback(reinterpret_cast<VirtualQCompleter::QCompleter_Event_Callback>(slot));
}
}

void QCompleter_Delete(QCompleter* self) {
    delete self;
}

File: binding/src/libqcompleter.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qcompleter.html
pub const qcompleter = struct {

    /// New constructs a new QCompleter object.
    pub fn New() QtC.QCompleter {
        return qtc.QCompleter_new();
    }


    /// New2 constructs a new QCompleter object.
    pub fn New2(model: ?*anyopaque) QtC.QCompleter {
        return qtc.QCompleter_new2(@ptrCast(model));
    }


    /// New3 constructs a new QCompleter object.
    pub fn New3(completions: []const u8) QtC.QCompleter {
        return qtc.QCompleter_new3(@ptrCast(completions));
    }


    /// New4 constructs a new QCompleter object.
    pub fn New4(parent: ?*anyopaque) QtC.QCompleter {
        return qtc.QCompleter_new4(@ptrCast(parent));
    }


    /// New5 constructs a new QCompleter object.
    pub fn New5(model: ?*anyopaque, parent: ?*anyopaque) QtC.QCompleter {
        return qtc.QCompleter_new5(@ptrCast(model), @ptrCast(parent));
    }


    /// New6 constructs a new QCompleter object.
    pub fn New6(completions: []const u8, parent: ?*anyopaque) QtC.QCompleter {
        return qtc.QCompleter_new6(@ptrCast(completions), @ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QCompleter_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qcompleter.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetWidget(self: ?*anyopaque, widget: ?*anyopaque) void {
        qtc.QCompleter_SetWidget(@ptrCast(self), @ptrCast(widget));
    }

    pub fn Widget(self: ?*anyopaque, ) QtC.QWidget {
        return qtc.QCompleter_Widget(@ptrCast(self));
    }

    pub fn SetModel(self: ?*anyopaque, c: ?*anyopaque) void {
        qtc.QCompleter_SetModel(@ptrCast(self), @ptrCast(c));
    }

    pub fn Model(self: ?*anyopaque, ) QtC.QAbstractItemModel {
        return qtc.QCompleter_Model(@ptrCast(self));
    }

    pub fn SetCompletionMode(self: ?*anyopaque, mode: i32) void {
        qtc.QCompleter_SetCompletionMode(@ptrCast(self), @intCast(mode));
    }

    pub fn CompletionMode(self: ?*anyopaque, ) i32 {
        return qtc.QCompleter_CompletionMode(@ptrCast(self));
    }

    pub fn SetFilterMode(self: ?*anyopaque, filterMode: i32) void {
        qtc.QCompleter_SetFilterMode(@ptrCast(self), @intCast(filterMode));
    }

    pub fn FilterMode(self: ?*anyopaque, ) i32 {
        return qtc.QCompleter_FilterMode(@ptrCast(self));
    }

    pub fn Popup(self: ?*anyopaque, ) QtC.QAbstractItemView {
        return qtc.QCompleter_Popup(@ptrCast(self));
    }

    pub fn SetPopup(self: ?*anyopaque, popup: ?*anyopaque) void {
        qtc.QCompleter_SetPopup(@ptrCast(self), @ptrCast(popup));
    }

    pub fn SetCaseSensitivity(self: ?*anyopaque, caseSensitivity: i32) void {
        qtc.QCompleter_SetCaseSensitivity(@ptrCast(self), @intCast(caseSensitivity));
    }

    pub fn CaseSensitivity(self: ?*anyopaque, ) i32 {
        return qtc.QCompleter_CaseSensitivity(@ptrCast(self));
    }

    pub fn SetModelSorting(self: ?*anyopaque, sorting: i32) void {
        qtc.QCompleter_SetModelSorting(@ptrCast(self), @intCast(sorting));
    }

    pub fn ModelSorting(self: ?*anyopaque, ) i32 {
        return qtc.QCompleter_ModelSorting(@ptrCast(self));
    }

    pub fn SetCompletionColumn(self: ?*anyopaque, column: i32) void {
        qtc.QCompleter_SetCompletionColumn(@ptrCast(self), column);
    }

    pub fn CompletionColumn(self: ?*anyopaque, ) i32 {
        return qtc.QCompleter_CompletionColumn(@ptrCast(self));
    }

    pub fn SetCompletionRole(self: ?*anyopaque, role: i32) void {
        qtc.QCompleter_SetCompletionRole(@ptrCast(self), role);
    }

    pub fn CompletionRole(self: ?*anyopaque, ) i32 {
        return qtc.QCompleter_CompletionRole(@ptrCast(self));
    }

    pub fn WrapAround(self: ?*anyopaque, ) bool {
        return qtc.QCompleter_WrapAround(@ptrCast(self));
    }

    pub fn MaxVisibleItems(self: ?*anyopaque, ) i32 {
        return qtc.QCompleter_MaxVisibleItems(@ptrCast(self));
    }

    pub fn SetMaxVisibleItems(self: ?*anyopaque, maxItems: i32) void {
        qtc.QCompleter_SetMaxVisibleItems(@ptrCast(self), maxItems);
    }

    pub fn CompletionCount(self: ?*anyopaque, ) i32 {
        return qtc.QCompleter_CompletionCount(@ptrCast(self));
    }

    pub fn SetCurrentRow(self: ?*anyopaque, row: i32) bool {
        return qtc.QCompleter_SetCurrentRow(@ptrCast(self), row);
    }

    pub fn CurrentRow(self: ?*anyopaque, ) i32 {
        return qtc.QCompleter_CurrentRow(@ptrCast(self));
    }

    pub fn CurrentIndex(self: ?*anyopaque, ) QtC.QModelIndex {
        return qtc.QCompleter_CurrentIndex(@ptrCast(self));
    }

    pub fn CurrentCompletion(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QCompleter_CurrentCompletion(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qcompleter.CurrentCompletion: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn CompletionModel(self: ?*anyopaque, ) QtC.QAbstractItemModel {
        return qtc.QCompleter_CompletionModel(@ptrCast(self));
    }

    pub fn CompletionPrefix(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QCompleter_CompletionPrefix(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qcompleter.CompletionPrefix: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetCompletionPrefix(self: ?*anyopaque, prefix: []const u8) void {
        const prefix_str = qtc.libqt_string{
    .len = prefix.len,
    .data = prefix.ptr,
};
qtc.QCompleter_SetCompletionPrefix(@ptrCast(self), prefix_str);
    }

    pub fn Complete(self: ?*anyopaque, ) void {
        qtc.QCompleter_Complete(@ptrCast(self));
    }

    pub fn SetWrapAround(self: ?*anyopaque, wrap: bool) void {
        qtc.QCompleter_SetWrapAround(@ptrCast(self), wrap);
    }

    pub fn PathFromIndex(self: ?*anyopaque, index: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QCompleter_PathFromIndex(@ptrCast(self), @ptrCast(index));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qcompleter.PathFromIndex: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Allows for overriding the related default method
    pub fn OnPathFromIndex(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) [*:0]const u8) void {
        qtc.QCompleter_OnPathFromIndex(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBasePathFromIndex(self: ?*anyopaque, index: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QCompleter_QBasePathFromIndex(@ptrCast(self), @ptrCast(index));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qcompleter.PathFromIndex: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SplitPath(self: ?*anyopaque, path: []const u8) []const u8 {
        const path_str = qtc.libqt_string{
    .len = path.len,
    .data = path.ptr,
};
return qtc.QCompleter_SplitPath(@ptrCast(self), path_str);
    }

    /// Allows for overriding the related default method
    pub fn OnSplitPath(self: ?*anyopaque, callback: *const fn (?*anyopaque, [*:0]const u8) callconv(.c) [*:0]const u8) void {
        qtc.QCompleter_OnSplitPath(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSplitPath(self: ?*anyopaque, path: []const u8) []const u8 {
        const path_str = qtc.libqt_string{
    .len = path.len,
    .data = path.ptr,
};
return qtc.QCompleter_QBaseSplitPath(@ptrCast(self), path_str);
    }

    pub fn EventFilter(self: ?*anyopaque, o: ?*anyopaque, e: ?*anyopaque) bool {
        return qtc.QCompleter_EventFilter(@ptrCast(self), @ptrCast(o), @ptrCast(e));
    }

    /// Allows for overriding the related default method
    pub fn OnEventFilter(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque, ?*anyopaque) callconv(.c) bool) void {
        qtc.QCompleter_OnEventFilter(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseEventFilter(self: ?*anyopaque, o: ?*anyopaque, e: ?*anyopaque) bool {
        return qtc.QCompleter_QBaseEventFilter(@ptrCast(self), @ptrCast(o), @ptrCast(e));
    }

    pub fn Event(self: ?*anyopaque, param1: ?*anyopaque) bool {
        return qtc.QCompleter_Event(@ptrCast(self), @ptrCast(param1));
    }

    /// Allows for overriding the related default method
    pub fn OnEvent(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) bool) void {
        qtc.QCompleter_OnEvent(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseEvent(self: ?*anyopaque, param1: ?*anyopaque) bool {
        return qtc.QCompleter_QBaseEvent(@ptrCast(self), @ptrCast(param1));
    }

    pub fn Activated(self: ?*anyopaque, text: []const u8) void {
        const text_str = qtc.libqt_string{
    .len = text.len,
    .data = text.ptr,
};
qtc.QCompleter_Activated(@ptrCast(self), text_str);
    }

    pub fn Activated2(self: ?*anyopaque, index: ?*anyopaque) void {
        qtc.QCompleter_Activated2(@ptrCast(self), @ptrCast(index));
    }

    pub fn Highlighted(self: ?*anyopaque, text: []const u8) void {
        const text_str = qtc.libqt_string{
    .len = text.len,
    .data = text.ptr,
};
qtc.QCompleter_Highlighted(@ptrCast(self), text_str);
    }

    pub fn Highlighted2(self: ?*anyopaque, index: ?*anyopaque) void {
        qtc.QCompleter_Highlighted2(@ptrCast(self), @ptrCast(index));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QCompleter_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qcompleter.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QCompleter_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qcompleter.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Complete1(self: ?*anyopaque, rect: ?*anyopaque) void {
        qtc.QCompleter_Complete1(@ptrCast(self), @ptrCast(rect));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QCompleter_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/completer.html#types
pub const enums = struct {
    pub const CompletionMode = enum {
        pub const PopupCompletion: i32 = 0;
        pub const UnfilteredPopupCompletion: i32 = 1;
        pub const InlineCompletion: i32 = 2;
    };

    pub const ModelSorting = enum {
        pub const UnsortedModel: i32 = 0;
        pub const CaseSensitivelySortedModel: i32 = 1;
        pub const CaseInsensitivelySortedModel: i32 = 2;
    };

};
